File: BEditor/BEditor.CLI/Range.cpp
#include "pch.h"

using namespace System;

using namespace BEditor::CLI::Media;
using Range_ = BEditor::CLI::Media::Range;

inline Range_::Range(int start, int end) {
	Start = start;
	End = end;
}

inline Range_ Range_::All::get() {
	return Range_(int::MinValue, int::MaxValue);
}
inline bool Range_::Equals(Range_ range) {
	return Start == range.Start && End == range.End;
}
inline bool Range_::Equals(Object^ obj) {
	return obj->GetType() == Range_::typeid && Equals((Range_)obj);
}
inline int Range_::GetHashCode() {
	return HashCode::Combine(Start, End);
}
inline String^ Range_::ToString() {
	return String::Format("(Start:{0} End:{1})", Start, End);
}

inline bool Range_::operator==(Range_ left, Range_ right) {
	return left.Equals(right);
}
inline bool Range_::operator!=(Range_ left, Range_ right) {
	return !left.Equals(right);
}

File: BEditor/BEditor.Core/Media/Range.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.Contracts;
using System.Runtime.InteropServices;
using System.Runtime.Serialization;
using System.Text;

namespace BEditor.Core.Media
{
#nullable enable
    /// <summary>
    /// 
    /// </summary>
    [Serializable]
    [StructLayout(LayoutKind.Sequential)]
    public struct Range : IEquatable<Range>
    {
        /// <summary>
        /// 
        /// </summary>
        public int Start { get; set; }

        /// <summary>
        /// 
        /// </summary>
        public int End { get; set; }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="start"></param>
        /// <param name="end"></param>
        public Range(int start, int end)
        {
            Start = start;
            End = end;
        }

        /// <summary>
        /// 
        /// </summary>
        public static Range All => new Range(int.MinValue, int.MaxValue);

        /// <inheritdoc />
        [Pure]
        public bool Equals(Range other)
        {
            return Start == other.Start && End == other.End;
        }

        /// <inheritdoc />
        [Pure]
        public override bool Equals(object? obj)
        {
            return obj is Range other && Equals(other);
        }

        /// <inheritdoc />
        [Pure]
        public override int GetHashCode()
        {
            unchecked
            {
                return (Start * 397) ^ End;
            }
        }

        /// <inheritdoc/>
        [Pure]
        public override string ToString() => $"(Start:{Start} End:{End})";


        /// <summary>
        /// 
        /// </summary>
        /// <param name="left"></param>
        /// <param name="right"></param>
        /// <returns></returns>
        [Pure]
        public static bool operator ==(Range left, Range right)
        {
            return left.Equals(right);
        }
        /// <summary>
        /// 
        /// </summary>
        /// <param name="left"></param>
        /// <param name="right"></param>
        /// <returns></returns>
        [Pure]
        public static bool operator !=(Range left, Range right)
        {
            return !(left == right);
        }
    }
}
